internal/controller/argocdcommitstatus_controller.go

```diff
@@ -0,0 +1,401 @@
+		}
+
+		logger.Error(err, "failed to get ArgoCDCommitStatus")
+		return ctrl.Result{}, fmt.Errorf("failed to get ArgoCDCommitStatus: %w", err)
+	}
+
+	ls, err := metav1.LabelSelectorAsSelector(argoCDCommitStatus.Spec.ApplicationSelector)
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("failed to parse label selector: %w", err)
+	}
+	// TODO: we should setup a field index and only list apps related to the currently reconciled app
+	var ulArgoCDApps unstructured.UnstructuredList
+	ulArgoCDApps.SetGroupVersionKind(gvk)
+	err = r.Client.List(ctx, &ulArgoCDApps, &client.ListOptions{
+		LabelSelector: ls,
+	})
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("failed to list CommitStatus objects: %w", err)
+	}
+
+	gitAuthProvider, repositoryRef, err := r.getGitAuthProvider(ctx, argoCDCommitStatus)
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("failed to get git auth provider: %w", err)
+	}
+
+	groupedArgoCDApps, err := r.groupArgoCDApplicationsWithPhase(&argoCDCommitStatus, ulArgoCDApps)
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("failed to get ArgoCDApplication: %w", err)
+	}
+
+	for targetBranch, appsInEnvironment := range groupedArgoCDApps {
+		gitOperation, err := git.NewGitOperations(ctx, r.Client, gitAuthProvider, r.PathLookup, repositoryRef, &argoCDCommitStatus, targetBranch)
+		if err != nil {
+			return ctrl.Result{}, fmt.Errorf("failed to initialize git client: %w", err)
+		}
+
+		var resolvedSha string
+		resolvedSha, err = gitOperation.LsRemote(ctx, targetBranch)
+		if err != nil {
+			return ctrl.Result{}, fmt.Errorf("failed to ls-remote sha: %w", err)
+		}
+
+		mostRecentLastTransitionTime := r.getMostRecentLastTransitionTime(appsInEnvironment)
+
+		resolvedPhase, desc := r.calculateAggregatedPhaseAndDescription(appsInEnvironment, resolvedSha, mostRecentLastTransitionTime)
+
+		err = r.updateAggregatedCommitStatus(ctx, argoCDCommitStatus, targetBranch, resolvedSha, resolvedPhase, desc)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+	}
+
+	err = r.Status().Update(ctx, &argoCDCommitStatus)
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("failed to update ArgoCDCommitStatus status: %w", err)
+	}
+
+	return ctrl.Result{RequeueAfter: 15 * time.Second}, nil // Timer for now :(
+}
+
+func (r *ArgoCDCommitStatusReconciler) groupArgoCDApplicationsWithPhase(argoCDCommitStatus *promoterv1alpha1.ArgoCDCommitStatus, ulAppList unstructured.UnstructuredList) (map[string][]*aggregate, error) {
+	aggregates := map[string][]*aggregate{}
+	argoCDCommitStatus.Status.ApplicationsSelected = []promoterv1alpha1.ApplicationsSelected{}
+	repo := ""
+
+	for _, ulApp := range ulAppList.Items {
+		var application argocd.ArgoCDApplication
+		err := runtime.DefaultUnstructuredConverter.FromUnstructured(ulApp.Object, &application)
+		if err != nil {
+			return map[string][]*aggregate{}, fmt.Errorf("failed to cast unstructured object to typed object: %w", err)
+		}
+
+		if application.Spec.SourceHydrator == nil {
+			continue
```

Comment:
Maybe just go ahead and throw an error? Would be consistent with how we handle non-matching repo URLs.